---
title: "18S ASV check"
author: "Denise Ong"
date: "3/10/2023"
output: pdf_document
---

This is to subset the ASVs with low bootstrap to check.

# Initialize
```{r, echo=TRUE, message=FALSE, warning=FALSE}
library(here)
source(here("phyloseq", "init_files", "init.R")) #for libraries
source(here("phyloseq", "init_files", "init_markdown.R")) #for markdown libraries
```

#read data
```{r}
asv<- readxl::read_excel(here("DADA2", "output_18SV4", "TAN1810_18SV4_metapr2_asv.xlsx")) %>%
    slice_head(n=500) %>%
    filter(species_boot <90) %>%
    filter(class != "Syndiniales") %>%
    filter(order != "Dinophyceae_X") %>%
    filter(!(division %in% c("Metazoa", "Fungi", "Pseudofungi"))) %>%
    filter(supergroup != "Opisthokonta") %>%
    filter(class != "Embryophyceae") %>%
    filter(genus != "Padina")
    
writexl::write_xlsx(asv, "TAN1810_18SV4_metapr2_asv_checks.xlsx")



    asv_sub <- asv %>%
      select(asv_code, sequence, kingdom:species)
    seq_out <- Biostrings::DNAStringSet(asv_sub$sequence)
    names(seq_out) <- str_c(asv_sub$asv_code,asv_sub$class, asv_sub$order, asv_sub$species, sep="|") #need to change.
    Biostrings::writeXStringSet(seq_out, "18sv4_check.fasta",
                                compress=FALSE, width = 20000)


```


ignore this
```{r}
asv_set47 <- readxl::read_excel(here("DADA2", "merge_filt_phyloseq", "metapr2_asv_set_47_Eukaryota copy.xlsx")) %>%
  select(-seq_name, -dataset_id, -gene, -asv_id)
asv_do <- readxl::read_excel(here("DADA2", "output_18SV4", "TAN1810_18SV4_metapr2_asv.xlsx")) %>%
   select(-seq_name, -dataset_id, -gene)

asv_merge <- full_join(asv_set47, asv_do)
#read merged phyloseq - asv set 47 processed by DV and dataset D4 processed by DO.
# contains all samples from TAN 1810 cruise from 2AM CTD filtered seawater samples, 18sv4 region.
ps_filt_sort<- read_rds(here("DADA2", "merge_filt_phyloseq","phyloseq_merge_asv_set_47_Eukaryota_D4_filt_sort_seqhash.RDS"))

junk <- data.frame(ps_filt_sort@sam_data)
junk<- data.frame(ps_filt_sort@otu_table)

# ctd number U1938 replace with U9138, U1941 replace with U9141.
sample_data(ps_filt_sort)$ctd_cast = str_replace(sample_data(ps_filt_sort)$ctd_cast, "U1938", "U9138") #filt
sample_data(ps_filt_sort)$ctd_cast = str_replace(sample_data(ps_filt_sort)$ctd_cast, "U1941", "U9141") #sort


# Remove ASVs that are not present in the dataset
ps_filt_sort <- ps_filt_sort %>% 
    phyloseq::filter_taxa(function(x) sum(x) > 0 , TRUE) 

## Filtration based on taxonomy - remove metazoa and fungi
ps_filt_sort<- ps_filt_sort %>% 
    subset_taxa(!(supergroup %in% c("Opisthokonta")))%>%
    subset_taxa(!(division %in% c("Pseudofungi")))%>%
    subset_taxa(!(division %in% c("Metazoa", "Fungi"))) %>%
    subset_taxa(!(class == "Embryophyceae")) # to remove the plants.Likely from contamination. Checked with Adriana and Daniel, ok to remove.
```

