---
title: "20210929_18S"
author: "Denise Ong"
date: "9/29/2021"
output: html_document
---
Basic plots for 18SV4 sorted Pico and Nano.

# Initialize
```{r, echo=TRUE, message=FALSE, warning=FALSE}

source("phyloseq/init.R") #for libraries
source("phyloseq/init_markdown.R") #for markdown libraries
source("phyloseq/define_functions.R") #functions to 1.transform to dataframe, 2.normalise and 3.create treemaps 
source("phyloseq/colours_18sv4.R")
```

### Function to extract abundant taxa (> fraction)
* Only take asvs that represent at least 10% of reads in any given sample

```{r}
ps_abund <- function(ps, fraction=0.10) {
  
  ps <- filter_taxa(ps, function(x) sum(x > total*fraction) > 0, TRUE)
}
```
# Sorted pico
## Read data
```{r}
source("phyloseq/read_pico.R")
# Do normalizations and transformation
ps_pico <- phyloseq_normalize_median(ps_pico)
long_pico <- phyloseq_transform_to_long(ps_pico)

ps_pico_2 <- ps_pico %>% 
    subset_taxa(!(supergroup %in% c("Opisthokonta")))%>%
    subset_taxa(!(division %in% c("Pseudofungi")))%>%
    subset_taxa(!(class %in% c("Syndiniales")) )
ps_pico_2 <- phyloseq_normalize_median(ps_pico_2)
long_pico_2 <- phyloseq_transform_to_long(ps_pico_2)

```

## General plot - by experiment and depth
```{r}
p1 <- long_pico %>%
  mutate(class = fct_infreq(class))%>%
 # mutate(class = fct_lump_prop(class, 0.10)) %>%
  mutate(class = fct_lump_n(class, 10, other_level = "Others")) %>%
  ggplot(aes(x=vial, y=n_reads, fill=class, color = class)) + 
        geom_bar(position="fill", stat="identity", size=0.5) + 
        theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1))+
        #facet_grid(fct_rev(SAMPLE)~vial_type, scales = "free_x")
        facet_grid(fct_rev(SAMPLE)~cycle_exp_location_name, scales = "free_x") +
        ggtitle("Sorted pico - class level") +
        scale_color_manual(values=colours_18s_class, limits = force) + 
        scale_fill_manual(values=colours_18s_class, limits = force) 
p1

#pdf("phyloseq/figures_20211021/sorted pico class_vial.pdf", height=5, width=8) ; plot(p1) ; dev.off()
```

## Average of light vials only, by cycle and depth
```{r}
p1 <- long_pico_2 %>%
  filter(vial_type == "light") %>%
  mutate(class = fct_lump_n(class, 8, other_level = "Others")) %>%
 # mutate(class = factor(class, levels = c("Mamiellophyceae", "Chloropicophyceae", "Bacillariophyta", "Bolidophyceae", "Chrysophyceae", "Pelagophyceae", "Syndiniales", "Dinophyceae", "Prymnesiophyceae","Choanoflagellatea", "Others"))) %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color = class, fill = class)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.border = element_blank(), 
              panel.spacing.x = unit(0,"lines"),
              panel.spacing.y = unit(0,"lines"),
              axis.line.y.left = element_line())+
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        ggtitle("Picoeukaryotes") +
        scale_color_manual(values=colours_18s_class, limits = force) + 
        scale_fill_manual(values=colours_18s_class, limits = force) +
  ylab("Percent of reads")
p1

#pdf("phyloseq/figures_ACOP/sorted pico class_light.pdf", height=5, width=5) ; plot(p1) ; dev.off()
```

## order level - Average of light vials only, by cycle and depth
```{r}
table_2 <- long_pico_2 %>%
  filter(vial_type == "light") %>%
  mutate(order = fct_lump_n(order, 10, other_level = "Others")) %>%
  select(division, class, order) %>%
  distinct() %>%
  filter(order != "Others") %>%
  arrange(division, class, order)

p1 <- long_pico_2 %>%
  filter(vial_type == "light") %>%
  mutate(order = fct_lump_n(order, 10, other_level = "Others")) %>%
  mutate(order = factor(order, levels = c("Chloropicales", "Mamiellales", "Prymnesiales", "Phaeocystales", "Prymnesiophyceae_Clade_D", "Bacillariophyta_X", "Parmales", "Chrysophyceae_X", "Dictyochophyceae_X","Pelagomonadales", "Others"))) %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color = order, fill = order)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.border = element_blank(), 
              legend.position = "none",
              panel.spacing.x = unit(0,"lines"),
              panel.spacing.y = unit(0,"lines"),
              axis.line.y.left = element_line())+
        #facet_grid(fct_rev(SAMPLE)~vial_type, scales = "free_x")
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        ggtitle("Picoeukaryotes") +
        scale_color_manual(values=colours_18s_order, limits = force) + 
        scale_fill_manual(values=colours_18s_order, limits = force) +
  ylab("Percent of reads")
p1

pdf("phyloseq/figures_ACOP/sorted pico order_light.pdf", height=6, width=3.5) ; plot(p1) ; dev.off()
```

## Top 15 species for each cycle
```{r}
for (i in c("ST1", "ST2", "SA1", "SA2", "SA3")) {
p <- long_pico_2 %>%
  filter(cycle_name_2 == i) %>%
  select(class, species, n_reads) %>%
  group_by(species) %>%
  mutate(n_reads = mean(n_reads)) %>% # sum of all reads or avg?
  distinct() %>%
  arrange(desc(n_reads)) %>%
  ungroup() %>%
  dplyr::slice_head(n = 10) %>%
  ggplot(aes(x=n_reads, y=reorder(species,n_reads), fill=class)) +
    geom_bar(stat="identity") + 
    ggtitle(str_c("Top 10 species Pico - ", i)) +
    scale_color_manual(values=colours_18s_class, limits = force) + 
    scale_fill_manual(values=colours_18s_class, limits = force) +
  theme(axis.text.y = element_text(angle = 45, hjust = 1, vjust = 0.5, size = 7))+
  xlab("Mean read abundance per sample") +
  ylab(NULL)
print(p)

pdf(str_c("phyloseq/figures_ACOP/sorted pico top 10_", i, ".pdf"), height=3, width=5) ; plot(p) ; dev.off()
}

```
## Look at different groups
```{r}
for(i in c("Prymnesiophyceae")) {
ps_pico_sub <- ps_pico %>% 
    subset_taxa(class == i)
ps_pico_sub <- phyloseq_normalize_median(ps_pico_sub)
long_pico_sub <- phyloseq_transform_to_long(ps_pico_sub)

p <- long_pico_sub %>%
  ggplot(aes(x=vial, y=n_reads, color=order, fill=order)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1)) +
        facet_grid(fct_rev(SAMPLE)~ cycle_exp_location_name , scales = "free_x") +
  ggtitle(i)
print(p)

p1 <- long_pico_sub %>%
  filter(vial_type == "light") %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color=order, fill=order)) + 
        geom_bar(position="fill", stat="identity") + 
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        #scale_color_manual(values=getPalette, limits = force) + 
        #scale_fill_manual(values=getPalette, limits = force) +
  ggtitle(i)
print(p1)
}

for(i in c("Pelagomonadales")) {
ps_pico_sub <- ps_pico %>% 
    subset_taxa(order == i)
ps_pico_sub <- phyloseq_normalize_median(ps_pico_sub)
long_pico_sub <- phyloseq_transform_to_long(ps_pico_sub)

p <- long_pico_sub %>%
  ggplot(aes(x=vial, y=n_reads, color=species, fill=species)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1)) +
        facet_grid(fct_rev(SAMPLE)~ cycle_exp_location_name , scales = "free_x") +
  ggtitle(i)
print(p)

p1 <- long_pico_sub %>%
  filter(vial_type == "light") %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color=species, fill=species)) + 
        geom_bar(position="fill", stat="identity") + 
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        #scale_color_manual(values=getPalette, limits = force) + 
        #scale_fill_manual(values=getPalette, limits = force) +
  ggtitle(i)
print(p1)
}
```

```{r}

## Filtration based on taxonomy - remove metazoa and fungi
ps_mam <- ps_pico %>% 
  subset_taxa(class == "Mamiellophyceae")

# Do normalizations and transformation
ps_mam <- phyloseq_normalize_median(ps_mam)
long_mam <- phyloseq_transform_to_long(ps_mam)

p1 <- long_mam %>%
  filter(vial_type == "light") %>%
  mutate(species = fct_lump_n(species, 4, other_level = "Others")) %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color = species, fill = species)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.border = element_blank(), 
              panel.spacing.x = unit(0,"lines"),
              panel.spacing.y = unit(0,"lines"),
              axis.line.y.left = element_line())+
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        ggtitle("Mamiellophyceae") +
  scale_fill_paletteer_d("ggthemes::Classic_20")+
  scale_colour_paletteer_d("ggthemes::Classic_20")+
  ylab("Percent of reads")
p1

pdf("phyloseq/figures_ACOP/sorted pico mam.pdf", height=5, width=5) ; plot(p1) ; dev.off()


```

# Sorted nano
## Read data
```{r}
source("phyloseq/read_nano.R")

# Do normalizations and transformation
ps_nano <- phyloseq_normalize_median(ps_nano)
long_nano <- phyloseq_transform_to_long(ps_nano)

ps_nano_2 <- ps_nano %>% 
    subset_taxa(!(supergroup %in% c("Opisthokonta")))%>%
    subset_taxa(!(division %in% c("Pseudofungi")))%>%
    subset_taxa(!(class %in% c("Syndiniales")) )
ps_nano_2 <- phyloseq_normalize_median(ps_nano_2)
long_nano_2 <- phyloseq_transform_to_long(ps_nano_2)

```

## General plot - by experiment and depth
```{r}
p <- long_nano_2 %>%
  mutate(class = fct_infreq(class))%>%
  #group_by(file_code, class) %>%
  #mutate(class = fct_lump_prop(class, 0.20)) %>%
  mutate(class = fct_lump_n(class, 20, other_level = "Others")) %>%
  ggplot(aes(x=vial, y=n_reads, color=class, fill=class)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1)) +
        facet_grid(fct_rev(SAMPLE)~ cycle_exp_location_name , scales = "free_x") +
        ggtitle("Sorted nano - class level") +
        scale_color_manual(values=colours_18s_class, limits = force) + 
        scale_fill_manual(values=colours_18s_class, limits = force) 
p
#pdf("phyloseq/figures_20211021/sorted pico class_vial.pdf", height=5, width=8) ; plot(p1) ; dev.off()
```
## Average of light vials only, by cycle and depth
```{r}
p1 <- long_nano %>%
  filter(vial_type == "light") %>%
  mutate(class = fct_infreq(class))%>%
  mutate(class = fct_lump_n(class, 10, other_level = "Others")) %>%
 # mutate(class = factor(class, levels = c("Mamiellophyceae", "Pyramimonadophyceae","Bacillariophyta","Dictyochophyceae","MOCH-2","Pelagophyceae",    "Syndiniales", "Dinophyceae", "Prymnesiophyceae","Cryptophyceae", "Others"))) %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color=class, fill=class)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.border = element_blank(), 
              panel.spacing.x = unit(0,"lines"),
              panel.spacing.y = unit(0,"lines"),
              axis.line.y.left = element_line())+
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        ggtitle("Nanoeukaryotes") +
        scale_color_manual(values=colours_18s_class, limits = force) + 
        scale_fill_manual(values=colours_18s_class, limits = force) +
  ylab("Percent of reads")
p1
#pdf("phyloseq/figures_ACOP/sorted nano class_light.pdf", height=5, width=5) ; plot(p1) ; dev.off()
```

## order level
```{r}
table <- long_nano_2 %>%
  filter(vial_type == "light") %>%
  mutate(order = fct_lump_n(order, 15, other_level = "Others")) %>%
  select(supergroup, division, class, order) %>%
  distinct() %>%
  filter(order != "Others") %>%
  arrange(supergroup, division, class, order)

p1 <- long_nano_2 %>%
  filter(vial_type == "light") %>%
#  mutate(class = fct_infreq(class))%>%
  mutate(order = fct_lump_n(order, 15, other_level = "Others")) %>%
  mutate(order = factor(order, levels = c("Dolichomastigales", "Pyramimonadales","Cryptomonadales", "Dinophyceae_X", "Suessiales", "Gymnodiniales","Peridiniales",    "Prorocentrales", "Isochrysidales", "Prymnesiales", "Phaeocystales","Syracosphaerales" , "Bacillariophyta_X", "Dictyochophyceae_X", "MOCH-2_X", "Others"))) %>%
    filter(sample_name != "nano-12") %>%
   # filter(sample_name != "syn-11") %>%
   # filter(sample_name != "syn-12") %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color=order, fill=order)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.border = element_blank(), 
              legend.position = "none",
              panel.spacing.x = unit(0,"lines"),
              panel.spacing.y = unit(0,"lines"),
              axis.line.y.left = element_line())+
        #facet_grid(fct_rev(SAMPLE)~vial_type, scales = "free_x")
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
        ggtitle("Nanoeukaryotes") +
        scale_color_manual(values=colours_18s_order) + 
        scale_fill_manual(values=colours_18s_order) +
  ylab("Percent of reads")
p1

pdf("phyloseq/figures_ACOP/sorted nano order_light.pdf", height=6, width=3.5) ; plot(p1) ; dev.off()
```

## Look at different groups
```{r}
for(i in c("Dinophyceae", "Prymnesiophyceae", "Cryptophyceae", "Bacillariophyta")) {
ps_nano_sub <- ps_nano %>% 
    subset_taxa(class == i)
ps_nano_sub <- phyloseq_normalize_median(ps_nano_sub)
long_nano_sub <- phyloseq_transform_to_long(ps_nano_sub)

p <- long_nano_sub %>%
  ggplot(aes(x=vial, y=n_reads, color=order, fill=order)) + 
        geom_bar(position="fill", stat="identity") + 
        theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1)) +
        facet_grid(fct_rev(SAMPLE)~ cycle_exp_location_name , scales = "free_x") +
  ggtitle(i)
print(p)

p1 <- long_nano_sub %>%
  filter(vial_type == "light") %>%
  ggplot(aes(x=cycle_name_2, y=n_reads, color=order, fill=order)) + 
        geom_bar(position="fill", stat="identity") + 
        facet_grid(fct_rev(SAMPLE)~factor(cycle_name_2, levels = c("ST1", "ST2", "SA1", "SA2", "SA3")), scales = "free_x") +
  ggtitle(i)
print(p1)
}
```

## Top 15 species - sum
```{r}
for (i in c("ST1", "ST2", "SA1", "SA2", "SA3")) {
p <- long_nano_2 %>%
  filter(sample_name != "nano-12") %>%
  filter(cycle_name_2 == i) %>%
  select(class, species, n_reads) %>%
  group_by(species) %>%
  mutate(n_reads = sum(n_reads)) %>% # sum of all reads or avg?
  distinct() %>%
  arrange(desc(n_reads)) %>%
  ungroup() %>%
  dplyr::slice_head(n = 15) %>%
  ggplot(aes(x=n_reads, y=reorder(species,n_reads), fill=class)) +
    geom_bar(stat="identity") + 
    ggtitle(str_c("Top 15 species Pico -", i)) +
    scale_color_manual(values=colours_18s_class, limits = force) + 
    scale_fill_manual(values=colours_18s_class, limits = force) +
  xlab("Number of reads") +
  ylab(NULL)
print(p)
}

```
## Top 15 species - average
```{r}
for (i in c("ST1", "ST2", "SA1", "SA2", "SA3")) {
p <- long_nano_2 %>%
  filter(cycle_name_2 == i) %>%
  select(class, species, n_reads) %>%
  group_by(species) %>%
  mutate(n_reads = mean(n_reads)) %>% # average
  distinct() %>%
  arrange(desc(n_reads)) %>%
  ungroup() %>%
  dplyr::slice_head(n = 10) %>%
  ggplot(aes(x=n_reads, y=reorder(species,n_reads), fill=class)) +
    geom_bar(stat="identity") + 
    ggtitle(str_c("Top 10 species Nano -", i)) +
    scale_color_manual(values=colours_18s_class, limits = force) + 
    scale_fill_manual(values=colours_18s_class, limits = force) +
  theme(axis.text.y = element_text(angle = 45, hjust = 1, vjust = 0.5, size = 7))+
  xlab("Mean read abundance per sample") +
  ylab(NULL)
print(p)

pdf(str_c("phyloseq/figures_ACOP/sorted nano top 10_", i, ".pdf"), height=3, width=5) ; plot(p) ; dev.off()
}

```
# CTD 18s
##Read data
```{r}
# Load the phyloseq files
ps_ctd_nested<- read_rds("output_18SV4/TAN1810_18SV4phyloseq_CTD_18SV4.RDS") 

#copy the column of interest to create new column, and replacing names for water masses SA and ST
sample_data(ps_ctd_nested)$water_mass <- sample_data(ps_ctd_nested)$cycle
sample_data(ps_ctd_nested)$water_mass = str_replace(sample_data(ps_ctd_nested)$water_mass, "1|2|5", "SA") 
sample_data(ps_ctd_nested)$water_mass = str_replace(sample_data(ps_ctd_nested)$water_mass, "3|4", "ST")

#copy the column of interest, creating new column and replace for cycle names. Not using this column for now. 
sample_data(ps_ctd_nested)$cycle_location <- sample_data(ps_ctd_nested)$cycle
sample_data(ps_ctd_nested)$cycle_location = str_replace(sample_data(ps_ctd_nested)$cycle_location, "1", "C")
sample_data(ps_ctd_nested)$cycle_location = str_replace(sample_data(ps_ctd_nested)$cycle_location, "2", "D") 
sample_data(ps_ctd_nested)$cycle_location = str_replace(sample_data(ps_ctd_nested)$cycle_location, "5", "E") 
sample_data(ps_ctd_nested)$cycle_location = str_replace(sample_data(ps_ctd_nested)$cycle_location, "3", "A")
sample_data(ps_ctd_nested)$cycle_location = str_replace(sample_data(ps_ctd_nested)$cycle_location, "4", "B")

sample_data(ps_ctd_nested)$cycle_name <- sample_data(ps_ctd_nested)$cycle
sample_data(ps_ctd_nested)$cycle_name = str_replace(sample_data(ps_ctd_nested)$cycle_name, "1", "SA-Cycle_1")
sample_data(ps_ctd_nested)$cycle_name = str_replace(sample_data(ps_ctd_nested)$cycle_name, "2", "SA-Cycle_2") 
sample_data(ps_ctd_nested)$cycle_name = str_replace(sample_data(ps_ctd_nested)$cycle_name, "5", "SA-Cycle_5") 
sample_data(ps_ctd_nested)$cycle_name = str_replace(sample_data(ps_ctd_nested)$cycle_name, "3", "ST-Cycle_3")
sample_data(ps_ctd_nested)$cycle_name = str_replace(sample_data(ps_ctd_nested)$cycle_name, "4", "ST-Cycle_4")

sample_data(ps_ctd_nested)$cycle_location_name <- str_c(sample_data(ps_ctd_nested)$cycle_location, sample_data(ps_ctd_nested)$cycle_name, sep ="_")

# Remove ASVs that are not present in the dataset
ps_ctd_nested <- ps_ctd_nested %>% 
    phyloseq::filter_taxa(function(x) sum(x) > 0 , TRUE) 
## Filtration based on taxonomy - remove metazoa and fungi
ps_ctd_nested<- ps_ctd_nested %>% 
      subset_taxa(!(division %in% c("Metazoa", "Fungi"))) %>%
      subset_taxa(!(class == "Embryophyceae")) # to remove the plants.Likely from contamination. Checked with Adriana and Daniel, ok to remove.

# Do normalizations and transformation
ps_ctd_nested <- phyloseq_normalize_median(ps_ctd_nested)
long_ctd_nested <- phyloseq_transform_to_long(ps_ctd_nested)
```

```{r}
p5 <- long_ctd_nested %>%
   # mutate(Subclade = fct_infreq(Subclade))%>%
   # filter(sample_name != "syn-negr1-2") %>%
  ggplot(aes(x=file_code, y=n_reads, color=division, fill=division)) + 
        geom_bar(position="fill", stat="identity") + 
  theme(text=element_text(size=10),
              axis.text.x = element_text(angle = 45, hjust = 1), 
              plot.title = element_text(size = 10)) + 
        facet_grid(.~cycle_location_name, scales = "free_x") 
       # ggtitle("CTD petB Denise primer-R1") 
       # scale_color_manual(values=divisionPalette, limits = force) + 
      #  scale_fill_manual(values=divisionPalette, limits = force)
p5

#pdf("phyloseq/figures_20210930/CTD Denise R1.pdf", height=5, width=8) ; plot(p) ; dev.off()
```

